presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecution.java

```diff
@@ -249,20 +249,24 @@ public void addSources(List<TaskSource> sources)
             // tell existing drivers about the new splits; it is safe to update drivers
             // multiple times and out of order because sources contain full record of
             // the unpartitioned splits
-            for (TaskSource source : updatedUnpartitionedSources.values()) {
-                // tell all the existing drivers this source is finished
-                for (WeakReference<Driver> driverReference : drivers) {
-                    Driver driver = driverReference.get();
-                    // the driver can be GCed due to a failure or a limit
-                    if (driver != null) {
-                        driver.updateSource(source);
-                    }
-                    else {
-                        // remove the weak reference from the list to avoid a memory leak
-                        // NOTE: this is a concurrent safe operation on a CopyOnWriteArrayList
-                        drivers.remove(driverReference);
-                    }
+            for (WeakReference<Driver> driverReference : drivers) {
+                Driver driver = driverReference.get();
+                // the driver can be GCed due to a failure or a limit
+                if (driver == null) {
+                    // remove the weak reference from the list to avoid a memory leak
+                    // NOTE: this is a concurrent safe operation on a CopyOnWriteArrayList
+                    drivers.remove(driverReference);
+                    continue;
                 }
+                Optional<PlanNodeId> sourceId = driver.getSourceId();
```

Comment:
If you want, I think this could be compacted with some `.map` and `.ifPresent` calls

Reply:
I see your point. I prefer keeping it as is. I don't have a strong opinion here. And I don't think you do either. I'm leaving it unchanged for now.